mls-rs/src/group_joiner.rs

```diff
@@ -0,0 +1,240 @@
+
+    // Needed for reinit
+    #[cfg(feature = "psk")]
+    additional_psk: Option<crate::psk::secret::PskSecretInput>,
+}
+
+impl<'a, 'b, C: ClientConfig> GroupJoiner<'a, 'b, C> {
+    // Info
+    pub fn required_external_psks(&self) -> impl Iterator<Item = &ExternalPskId> {
+        self.group_secrets
+            .psks
+            .iter()
+            .filter_map(|psk| match &psk.key_id {
+                JustPreSharedKeyID::External(psk) => Some(psk),
+                _ => None,
+            })
+    }
+
+    pub fn required_resumption_psks(&self) -> impl Iterator<Item = &ResumptionPsk> {
+        self.group_secrets
+            .psks
+            .iter()
+            .filter_map(|psk| match &psk.key_id {
+                JustPreSharedKeyID::Resumption(psk) => Some(psk),
+                _ => None,
+            })
+    }
+
+    pub fn cipher_suite(&self) -> CipherSuite {
+        self.welcome.cipher_suite
+    }
+
+    // Settings
+    pub fn ratchet_tree(self, tree: ExportedTree<'b>) -> Self {
+        Self {
+            tree: Some(tree),
+            ..self
+        }
+    }
+
+    pub fn signature_secret_key(self, signature_secret_key: SignatureSecretKey) -> Self {
+        Self {
+            signer: Some(signature_secret_key),
+            ..self
+        }
+    }
+
+    // TODO with_psks
+
+    // Reinit
+    #[cfg(feature = "psk")]
+    pub(crate) fn additional_psk(self, additional_psk: crate::psk::secret::PskSecretInput) -> Self {
+        Self {
+            additional_psk: Some(additional_psk),
+            ..self
+        }
+    }
+
+    // Joining
+    #[cfg_attr(not(mls_build_async), maybe_async::must_be_sync)]
+    pub(crate) async fn new(
+        config: C,
+        welcome_msg: &'a MlsMessage,
+        key_package_data: KeyPackageData,
+        signer: Option<SignatureSecretKey>,
+    ) -> Result<Self, MlsError> {
+        let key_package = KeyPackage::mls_decode(&mut &*key_package_data.key_package_bytes)?;
+        let init_key = &key_package_data.init_key;
+
+        let (group_secrets, welcome) =
+            Group::decrypt_group_secrets(welcome_msg, &config, &key_package, init_key).await?;
+
+        Ok(Self {
+            group_secrets,
+            welcome,
+            tree: None,
+            config,
+            version: welcome_msg.version,
+            key_package,
+            leaf_secret: key_package_data.leaf_node_key,
+            signer,
+            #[cfg(feature = "psk")]
+            additional_psk: None,
+        })
+    }
+
+    /// Decrypt and validate the GroupInfo from this Welcome message. This requires
+    /// that the ratchet tree is a part of GroupInfo or has been provided with
+    /// [GroupJoiner::tree], and that all required PSKs have been provided [TODO].
+    #[cfg_attr(not(mls_build_async), maybe_async::must_be_sync)]
+    pub(crate) async fn decrypt_group_info(
+        &mut self,
+    ) -> Result<(GroupInfo, TreeKemPublic, PskSecret), MlsError> {
```

Comment:
[nit] Maybe cleaner to have DecryptedGroupInfo struct with these as fields?

Reply:
Done